Improve testing, add manual (slow) test of many packages

itsayellow opened the pull request:
<!-- add an 'x' in the brackets below -->
* [ ] I have added an entry to `docs/changelog.md`

## Summary of changes
This was originally part of a branch converting `venv_metadata_inspector.py` to use modern libraries.  To verify I had the new `venv_metadata_inspector.py` working I wanted to test it across MANY packages.  After everything was working I split this new testing capability into its own branch and PR to make it easier to parse.

Notable changes:
* Add `package_info.py` to `tests/` which contains dict `PKG` with "package name" keys, and containing data on each package, which can be used by tests
  * `'spec'` sub-key contains a standard pipx testing version for each package for more deterministic tests
  * other sub-keys contain information about the package such as apps, and apps of dependencies
* Add `test_install_all_packages.py` which is only run manually (`nox -s test_all_packages`), and tests pipx installation of 86 different packages, both with and without dependencies.
  * Output is not the usual pytest PASS or FAIL, but text files reporting on which packages: installed with the pipx tests' cleared system PATH, installed with the original PATH, or didn't install successfully with either.
* Add another Github workflow triggered via `workflow_dispatch` to exercise the above testing of 86 packages.  This can be activated only through the Github web UI, and does not run automatically.
  * All separate OS's reports are combined into a master report and provided as an Artifact for download.
* Modified the `prebuild_wheels` function in `noxfile.py` to hopefully be cleaner.  It also now expects a list of each end package to build all dependency wheels for, not specific dependency wheels.  This should make specifying wheels easier as you don't have to track down each specific dependency that is problematic.


## Test plan
<!-- provide evidence of testing, preferably with command(s) that can be copy+pasted by others -->
Tested by both CI tests and also by running
```
nox -s test_all_packages
```

Outputs files:
* `reports/all_packages_nodeps_report_darwin_3.9_20201212.txt`
* `reports/all_packages_nodeps_errors_darwin_3.9_20201212.txt`
* `reports/all_packages_deps_report_darwin_3.9_20201212.txt`
* `reports/all_packages_deps_errors_darwin_3.9_20201212.txt`

Example `reports/all_packages_nodeps_report_darwin_3.9_20201212.txt`:
<details>

```
========================================================================
darwin          Python 3.9      2020-12-12 22:36:26

package_spec            cleared PATH    system PATH     install time    
------------------------------------------------------------------------
ansible==2.9.13         PASS                             75s 
awscli==1.18.168        PASS                             24s 
b2==2.0.2               PASS                             15s 
beancount==2.3.3        PASS                             20s 
beets==1.4.9            PASS                             20s 
black==20.8b1           PASS                             15s 
cactus==3.3.3           PASS                             18s 
chert==19.1.0           PASS                             19s 
cloudtoken==0.1.707     PASS                             26s 
coala==0.11.0           PASS                             16s 
cookiecutter==1.7.2     PASS                             13s 
cython==0.29.21         PASS                             10s 
datasette==0.50.2       PASS                             21s 
diffoscope==154         PASS                             14s 
doc2dash==2.3.0         PASS                             22s 
doitlive==4.3.0         PASS                             12s 
gdbgui==0.14.0.1        FAIL            PASS             25s 156s
gns3-gui==2.2.15        PASS                             19s 
grow==1.0.0a10          PASS                             35s 
guake==3.7.0            PASS                              9s 
gunicorn==20.0.4        PASS                             10s 
howdoi==2.0.7           PASS                             16s 
httpie==2.3.0           PASS                             12s 
hyde==0.8.9             FAIL            FAIL             16s   3s
ipython==7.16.1         PASS                             16s 
isort==5.6.4            PASS                             10s 
jaraco.financial==2.0   FAIL            FAIL             16s   7s
kaggle==1.5.9           PASS                             14s 
kibitzr==6.0.0          PASS                             26s 
klaus==1.5.2            PASS                             17s 
kolibri==0.14.3         PASS                             31s 
Lektor==3.2.0           PASS                             21s 
localstack==0.12.1      PASS                             22s 
mackup==0.8.29          PASS                             13s 
magic-wormhole==0.12.0  PASS                             31s 
mayan-edms==3.5.2       FAIL            FAIL             78s 164s
mkdocs==1.1.2           PASS                             27s 
mycli==1.22.2           PASS                             16s 
nikola==8.1.1           PASS                             25s 
nox==2020.8.22          PASS                             12s 
pelican==4.5.0          PASS                             16s 
platformio==5.0.1       PASS                             14s 
ppci==0.5.8             PASS                             11s 
prosopopee==1.1.3       PASS                             17s 
ptpython==3.0.7         PASS                             13s 
pycowsay==0.0.0.1       PASS                              9s 
pylint==2.3.1           PASS                             17s 
ReText==7.1.0           PASS                             17s 
robotframework==3.2.2   PASS                             10s 
shell-functools==0.3.0  PASS                             10s 
speedtest-cli==2.1.2    PASS                              9s 
Sphinx==3.2.1           PASS                             18s 
sqlmap==1.4.10          PASS                             15s 
streamlink==1.7.0       PASS                             33s 
taguette==0.9.2         PASS                             18s 
term2048==0.2.7         PASS                             10s 
tox-ini-fmt==0.5.0      PASS                             11s 
visidata==2.0.1         PASS                             11s 
vulture==2.1            PASS                              9s 
Weblate==4.3.1          FAIL            PASS             35s 133s
youtube-dl==2020.9.20   PASS                             12s 
ZEO==5.2.2              PASS                             62s 

Summary
------------------------------------------------------------------------
gdbgui==0.14.0.1 needs prebuilt wheel
hyde==0.8.9 FAILS
jaraco.financial==2.0 FAILS
mayan-edms==3.5.2 FAILS
Weblate==4.3.1 needs prebuilt wheel

Finished 2020-12-12 23:05:55
Elapsed: 0:29:28
```

</details>

gaborbernat reviewed (approved):
![](https://media.tenor.co/images/d8c94a556aa20a3f58b843bef75e1f83/tenor.gif)

itsayellow commented:
Added some `assert`s so that packages that completely fail are shown in pytest output, and modified the nox session and github workflow to ensure that tests keep going regardless of any failures.
